newrelic_python_agent/plugins/mysql.py

```diff
@@ -0,0 +1,922 @@
+
+Like all other plugins, the config can be a dict, which will represent the config for a single instance.
+It can also be a list of instances, each with it's own set of settings.  The possible settings are:
+
+    name: A descriptive name for this instance, which is used as the name reported to newrelic.
+        type: string
+        default: None
+    metrics: The set of metric categories to collect
+        type: string or list(string)
+        default: "status,newrelic"
+        allowed values:
+            - "all" -- this will select all categories
+            - a comma-separated string of categories (e.g. status,newrelic,master,innodb_metrics)
+            - list of categories
+
+The following settings are passed to sql.connect():
+
+    host: The hostname or IP to connect to.
+        type: string
+        default: localhost
+    port: The port to connect to the database.
+        type: integer
+        default: 3306
+    user: The username to connect to mysql as.
+        type: string
+        default: newrelic
+    password: The password to connect to mysql as.
+        type: string
+        default: f63c225f4abe9e13
+    database: Database to use, None to not use a particular one.
+        type: string
+        default: None
+    connect_timeout: The timeout when connecting to the database (seconds).
+        type: integer
+        default: 5
+
+Any other settings specified will be passed directly to connect() (from pymsql or mysql.connect)
+
+Example:
+    mysql:
+      - name: localhost
+        host: localhost
+        user: monitor
+        password: monitor
+        metrics: status,newrelic,master,innodb_status
+      - name: db1 (master)
+        host: db1
+        port: 5432
+        metrics: all
+      - name: db1 (slave)
+        host: db1-slave
+        metrics:
+          - status
+          - newrelic
+          - slave
+          - master
+          - innodb_status
+          - innodb_mutex
+          - innodb_metrics
+
+"""
+
+import re
+import time
+import logging
+
+# Support both PyMySQL and mysql-connector-python interfaces
+try:
+    import pymysql as sql
+
+    def _errno(err):
+        """Extract error code from pymysql.Error"""
+        return err.args[0]
+    from pymysql.constants.ER import ACCESS_DENIED_ERROR as ER_ACCESS_DENIED_ERROR, BAD_DB_ERROR as ER_BAD_DB_ERROR
+except ImportError as e1:
+    import mysql.connector as sql
+
+    def _errno(err):
+        """Extract error code from mysql.connector.Error"""
+        return err.errno
+    from mysql.connector.errorcode import ER_ACCESS_DENIED_ERROR, ER_BAD_DB_ERROR
+
+from newrelic_python_agent.plugins import base
+
+LOGGER = logging.getLogger(__name__)
+
+DEFAULT_UNIT = "Operations"
+DEFAULT_TYPE = "gauge"
+DEFAULT_METRICS = ['status', 'newrelic']
+DEFAULT_CONNECT_ARGS = {
+    "port": 3306,
+    "database": None,
+    "host": "localhost",
+    "user": "newrelic",
+    "password": "f63c225f4abe9e13",
```

Comment:
Is this the default from the Java plugin? Can we default to no password?

Reply:
It seems to be the default for the Java plugin for the last 4 years: https://github.com/newrelic-platform/newrelic_mysql_java_plugin/blob/0c6db973c70f9227c074bc489d5d38ad77adf517/src/com/newrelic/plugins/mysql/instance/MySQLAgent.java#L34